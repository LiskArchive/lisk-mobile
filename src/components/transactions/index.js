import React, { Fragment } from 'react';
import { View, Animated } from 'react-native';
import List from './list';
import Footer from './footer';
import { H3, Small, A } from '../toolBox/typography';
import styles from './styles';
import { fromRawLsk } from '../../utilities/conversions';
import Icon from '../toolBox/icon';
import colors from '../../constants/styleGuide/colors';
import easing from '../../utilities/easing';


/**
 * The container component containing sign in and create account functionality
 *
 * @todo
 */
class Transactions extends React.Component {
  state = {
    initialAnimations: {
      opacity: new Animated.Value(0),
      top: new Animated.Value(20),
    },
  }

  componentDidMount() {
    this.initialFadeIn();
  }

  initialFadeIn = () => {
    const { opacity, top } = this.state.initialAnimations;
    Animated.timing(opacity, {
      toValue: 1,
      duration: 400,
      delay: 100,
    }).start();
    Animated.timing(top, {
      toValue: 0,
      duration: 400,
      delay: 100,
      easing: easing.easeInOutQuart,
    }).start();
  }

  onPress = () => {
    this.props.navigate('Send', { initialize: true });
  }

  render() {
    const {
      transactions, navigate, account, footer,
    } = this.props;
    const balance = account ? parseFloat(fromRawLsk(account.balance)) : '';
    const Anim = Animated.View;
    const { opacity, top } = this.state.initialAnimations;

    return (<Anim style={[styles.container, { opacity, top }]}>
      {
        (!transactions ||
          (transactions.confirmed.length === 0 && transactions.pending.length === 0)) ?
          <Fragment></Fragment> :
          <Fragment>
            <H3 style={styles.title}>Activity</H3>
            {!account.initialized && balance >= 0.2 ?
              <View style={styles.initContainer}>
                <Icon name='warning' color={colors.action1} size={18} />
                <Small style={styles.initText}>Your account is not initialized.
                  <A
                    style={styles.link}
                    onPress={this.onPress}> Initialize it now</A>
                </Small>
              </View> : null
            }
            <List
              navigate={navigate}
              account={account ? account.address : ''}
              pending={transactions.pending}
              transactions={transactions.confirmed} />
            {
              footer ? <Footer /> : null
            }
          </Fragment>
      }
    </Anim>);
  }
}

export default Transactions;
