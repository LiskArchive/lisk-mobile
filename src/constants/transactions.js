import { themes } from './styleGuide';

import accountInitializationLight from '../assets/images/txDetail/account-initialization-light.png';
import accountInitializationDark from '../assets/images/txDetail/account-initialization-dark.png';

import setSecondPassphraseLight from '../assets/images/txDetail/second-passphrase-light.png';
import setSecondPassphraseDark from '../assets/images/txDetail/second-passphrase-dark.png';

import registerDelegateLight from '../assets/images/txDetail/delegate-registration-light.png';
import registerDelegateDark from '../assets/images/txDetail/delegate-registration-dark.png';

import voteLight from '../assets/images/txDetail/vote-light.png';
import voteDark from '../assets/images/txDetail/vote-dark.png';

/**
 * Since react-navigation doesn't support i18n
 * I've created this dummy function to help i18n scanner
 * understand about these titles.
 * We can remove this as soon as react-navigation supports i18n or
 * we change the router to another lib with i18n support.
 *
 * @param {String} str
 * @returns {String} same as the input string
 */
const t = str => str;

export default {
  send: {
    type: 0,
    fee: 1e7,
    title: t('Transfer'),
    image: () => null,
  },
  accountInitialization: {
    type: 0,
    fee: 1e7,
    title: t('Account initialization'),
    image: theme =>
      theme === themes.light
        ? accountInitializationLight
        : accountInitializationDark,
  },
  setSecondPassphrase: {
    type: 1,
    fee: 5e8,
    title: t('Second passphrase registration'),
    image: theme =>
      theme === themes.light
        ? setSecondPassphraseLight
        : setSecondPassphraseDark,
  },
  registerDelegate: {
    type: 2,
    fee: 25e8,
    title: t('Delegate registration'),
    image: theme =>
      theme === themes.light ? registerDelegateLight : registerDelegateDark,
  },
  vote: {
    type: 3,
    fee: 1e8,
    title: t('Vote'),
    image: theme => (theme === themes.light ? voteLight : voteDark),
  },
};
