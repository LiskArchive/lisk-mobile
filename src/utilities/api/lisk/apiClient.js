import { APIClient } from '@liskhq/lisk-client';
import config from '../../../../lsk.config';

const createLiskAPIClient = () => {
  if (config.isTestnet) {
    return APIClient.createTestnetAPIClient({ node: config.nodeURL });
  }

  return APIClient.createMainnetAPIClient();
};

export default createLiskAPIClient();
