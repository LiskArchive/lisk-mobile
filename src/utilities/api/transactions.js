import getMappedFunction from './functionMapper';

/**
 * @typedef {Object} Transaction
 * @property {String} id ID or hash of the transaction.
 * @property {String} senderAddress
 * @property {String} recipientAddress
 * @property {String} amount
 * @property {String} fee
 * @property {Number} timestamp
 * @property {Number} confirmations
 * @property {Number} type transaction type for LSK (0 for BTC)
 * @property {String} data transaction data field for LSK (empty string for BTC)
 */

/**
 * @typedef {Object} TransactionsResponse
 * @property {Array.<Transaction>} data
 * @property {Object} meta
 * @property {Number} meta.count total number of transactions available on the API.
 * @property {Number} meta.offset
 * @property {Number} meta.limit
 */

/**
 * Retrieves transactions for related token with respect to the given payload.
 * @param {String} tokenType
 * @param {Object} data
 * @param {String} data.id ID or hash of a specific transaction.
 * @param {String} data.address
 * @param {Number} data.limit
 * @param {Number} data.offset
 * @returns {Promise<TransactionsResponse>}
 */
const get = (tokenType, data) =>
  getMappedFunction(tokenType, 'transactions', 'get')(data);

/**
 * Creates a raw, ready-to-broadcast transaction data with given payload.
 * @param {String} tokenType
 * @param {Object} data
 * @param {String} data.passphrase
 * @param {String} data.recipientAddress
 * @param {String} data.amount
 * @param {String} data.secondPassphrase - required if registered, only used for LSK.
 * @param {String} data.data - custom data field, only used for LSK.
 * @param {Number} data.dynamicFeePerByte - rate of miner/dynamic fee per byte, only used for BTC.
 * @returns {Promise}
 */
const create = (tokenType, data) =>
  getMappedFunction(tokenType, 'transactions', 'create')(data);

/**
 * Broadcasts a transaction to the specified token's blockchain.
 * @param {String} tokenType
 * @param {Object} transaction
 * @returns {Promise}
 */
const broadcast = (tokenType, transaction) =>
  getMappedFunction(tokenType, 'transactions', 'broadcast')(transaction);

export default {
  get,
  create,
  broadcast,
};
