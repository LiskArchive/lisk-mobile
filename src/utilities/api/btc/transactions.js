import bitcoin from 'bitcoinjs-lib';
import config from '../../../../btc.config';
import { extractAddress, getDerivedPathFromPassphrase } from './account';
import { merge } from '../../helpers';
import { validateAddress } from '../../validators';
import { tokenMap } from '../../../constants/tokens';

/**
 * Normalizes transaction data retrieved from Blockchain.info API
 * @param {Object} data
 * @param {String} data.address Base address to use for formatting transactions
 * @param {Array} data.list Transaction list retrieved from API
 */
const normalizeTransactionsResponse = ({ address, list }) =>
  list.map(({ tx, feeSatoshi, confirmations, timestamp }) => {
    const data = {
      id: tx.txid,
      timestamp: timestamp ? Number(timestamp) * 1000 : null,
      confirmations: confirmations || 0,
      type: 0,
      data: '',
    };

    data.fee = feeSatoshi;

    const ownedInput = tx.inputs.find(i =>
      i.txDetail.scriptPubKey.addresses.includes(address)
    );

    if (ownedInput) {
      data.senderAddress = address;
      const extractedAddress = tx.outputs[0].scriptPubKey.addresses[0];
      data.recipientAddress =
        validateAddress(tokenMap.BTC.key, extractedAddress) === 0
          ? extractedAddress
          : 'Unparsed Address';
      data.amount = tx.outputs[0].satoshi;
    } else {
      const output = tx.outputs.find(o =>
        o.scriptPubKey.addresses.includes(address)
      );
      const extractedAddress = tx.inputs[0].txDetail.scriptPubKey.addresses[0];
      data.senderAddress =
        validateAddress(tokenMap.BTC.key, extractedAddress) === 0
          ? extractedAddress
          : 'Unparsed Address';
      data.recipientAddress = address;
      data.amount = output.satoshi;
    }

    return data;
  });

export const get = ({ id, address, limit = 20, offset = 0 }) =>
  new Promise(async (resolve, reject) => {
    try {
      let response;

      if (id) {
        response = await fetch(
          `${config.url}/transaction/${id}`,
          config.requestOptions
        );
      } else {
        response = await fetch(
          `${config.url}/transactions/${address}?limit=${limit}&offset=${offset}&sort=height:desc`,
          config.requestOptions
        );
      }

      const json = await response.json();

      if (response.ok) {
        const data = normalizeTransactionsResponse({
          address,
          list: id ? [json.data] : json.data,
        });

        resolve({
          data,
          meta: json.meta ? { count: json.meta.total } : {},
        });
      } else {
        reject(json);
      }
    } catch (error) {
      reject(error);
    }
  });

/**
 * Normalizes transaction data retrieved from Blockchain.info API
 * @param {Object} data
 * @param {Number} data.inputCount
 * @param {Number} data.outputCount
 * @param {Number} data.dynamicFeePerByte - in satoshis/byte.
 */
export const calculateTransactionFee = ({
  inputCount,
  outputCount,
  dynamicFeePerByte,
}) =>
  (inputCount * 180 + outputCount * 34 + 10 + inputCount) * dynamicFeePerByte;

/**
 * Retrieves unspent tx outputs of a BTC address from Blockchain.info API
 * @param {String} address
 * @returns {Promise<Array>}
 */
export const getUnspentTransactionOutputs = address =>
  new Promise(async (resolve, reject) => {
    try {
      const response = await fetch(`${config.url}/utxo/${address}`);
      const json = await response.json();

      if (response.ok) {
        resolve(json.data);
      } else {
        reject(json);
      }
    } catch (error) {
      reject(error);
    }
  });

export const create = ({
  passphrase,
  recipientAddress,
  amount,
  dynamicFeePerByte,
}) =>
  new Promise(async (resolve, reject) => {
    try {
      amount = Number(amount);
      dynamicFeePerByte = Number(dynamicFeePerByte);

      const senderAddress = extractAddress(passphrase);
      const unspentTxOuts = await exports.getUnspentTransactionOutputs(
        senderAddress
      );

      // Estimate total cost (currently estimates max cost by assuming the worst case)
      const estimatedMinerFee = calculateTransactionFee({
        inputCount: unspentTxOuts.length,
        outputCount: 2,
        dynamicFeePerByte,
      });

      const estimatedTotal = amount + estimatedMinerFee;

      // Check if balance is sufficient
      const unspentTxOutsTotal = unspentTxOuts.reduce((total, tx) => {
        total += tx.value;
        return total;
      }, 0);

      if (unspentTxOutsTotal < estimatedTotal) {
        reject(new Error('Insufficient (estimated) balance'));
      }

      // Find unspent txOuts to spend for this tx
      let txOutIndex = 0;
      let sumOfConsumedOutputs = 0;
      const txOutsToConsume = [];

      while (sumOfConsumedOutputs <= estimatedTotal) {
        const tx = unspentTxOuts[txOutIndex];
        txOutsToConsume.push(tx);
        txOutIndex += 1;
        sumOfConsumedOutputs += tx.value;
      }

      const txb = new bitcoin.TransactionBuilder(config.network);

      // Add inputs from unspent txOuts
      // eslint-disable-next-line
      for (const tx of txOutsToConsume) {
        txb.addInput(tx.tx_hash, tx.tx_pos);
      }

      // Output to Recipient
      txb.addOutput(recipientAddress, amount);

      // Calculate final fee
      const calculatedMinerFee = calculateTransactionFee({
        inputCount: txOutsToConsume.length,
        outputCount: 2,
        dynamicFeePerByte,
      });

      // Calculate total
      const calculatedTotal = amount + calculatedMinerFee;

      // Output to Change Address
      const change = sumOfConsumedOutputs - calculatedTotal;
      txb.addOutput(senderAddress, change);

      // Sign inputs
      const derivedPath = getDerivedPathFromPassphrase(passphrase);
      const keyPair = bitcoin.ECPair.fromWIF(
        derivedPath.toWIF(),
        config.network
      );
      for (let i = 0; i < txOutsToConsume.length; i++) {
        txb.sign(i, keyPair);
      }

      resolve(txb.build().toHex());
    } catch (error) {
      reject(error);
    }
  });

export const broadcast = transactionHex =>
  new Promise(async (resolve, reject) => {
    try {
      const response = await fetch(
        `${config.url}/transaction`,
        merge(config.requestOptions, {
          method: 'POST',
          body: JSON.stringify({ tx: transactionHex }),
        })
      );

      const json = await response.json();

      if (response.ok) {
        resolve(json);
      } else {
        reject(json);
      }
    } catch (error) {
      reject(error);
    }
  });

/**
 * Generates a Transaction Explorer URL for given transaction id
 * based on the configured network type
 * @param {String} - Transaction ID
 * @returns {String} - URL
 */
export const getTransactionExplorerURL = id =>
  `${config.transactionExplorerURL}/${id}`;
