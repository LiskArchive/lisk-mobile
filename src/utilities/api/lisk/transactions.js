import { Platform } from 'react-native';
import * as Lisk from '@liskhq/lisk-client';
import {
  isTransfer,
  moduleAssetNameIdMap,
  transferAssetSchema
} from 'modules/Transactions/constants';
import * as LiskAndroidPatch from './utils/index';
import { apiClient } from './apiClient';
import config from '../../../../lsk.config';

const getAmount = (tx) => {
  if (moduleAssetNameIdMap.transfer === tx.moduleAssetId) {
    return tx.asset.amount;
  }
  if (moduleAssetNameIdMap.unlockToken === tx.moduleAssetId) {
    // eslint-disable-next-line no-undef
    return tx.asset?.unlockObjects?.reduce?.((a, b) => a + BigInt(b.amount), BigInt(0)).toString();
  }
  return 0;
};

/**
 * Normalizes transaction data retrieved from Lisk Core API
 * https://lisk.com/documentation/lisk-core/reference/api.html#/Transactions/get_transactions__id_
 */
const normalizeTransactionsResponse = (list, block) =>
  list.map((tx) => ({
    id: tx.id,
    senderAddress: tx.sender.address,
    recipientAddress: isTransfer({ moduleAssetId: tx.moduleAssetId })
      ? tx.asset.recipient.address
      : tx.sender.address,
    amount: getAmount(tx),
    fee: tx.fee,
    timestamp: tx.block?.timestamp,
    confirmations: block ? block.height - tx.block?.height + 1 : 0,
    nonce: tx.nonce,
    type: tx.moduleAssetName,
    moduleAssetId: tx.moduleAssetId,
    moduleAssetName: tx.moduleAssetName,
    data: tx.moduleAssetName === 'token:transfer' ? tx.asset.data : '',
    votes: tx.moduleAssetName === 'dpos:voteDelegate' ? tx.asset.votes : [],
    delegate: tx.moduleAssetName === 'dpos:registerDelegate' ? tx.asset.username : '',
    blockId: tx.block?.id,
    blockHeight: tx.block?.height,
  }));

export const get = async ({
  id, address, limit, offset
}) => {
  const block = await apiClient.getLatestBlock();
  if (id !== undefined) {
    const txs = await apiClient.getTransaction(id);
    return { data: normalizeTransactionsResponse(txs, block), meta: {} };
  }
  const { data, meta } = await apiClient.getTransactions(address, limit, offset);
  if (data) {
    return {
      data: normalizeTransactionsResponse(data, block),
      meta: {
        limit,
        offset,
        count: meta.total
      }
    };
  }
  return {
    data: [],
    meta: {}
  };
};

export const create = async ({
  nonce, passphrase, recipientAddress, amount, fee, reference
}) => {
  const {
    publicKey: senderPublicKey
  } = Lisk.cryptography.address.getAddressAndPublicKeyFromPassphrase(passphrase);
  const recipient = Lisk.cryptography.address.getAddressFromLisk32Address(recipientAddress);
  const tx = {
    moduleID: 2,
    assetID: 0,
    senderPublicKey,
    // eslint-disable-next-line no-undef
    fee: BigInt(fee),
    // eslint-disable-next-line no-undef
    nonce: BigInt(nonce),
    asset: {
      // eslint-disable-next-line no-undef
      amount: BigInt(amount),
      data: reference || '',
      recipientAddress: recipient
    },
    signatures: []
  };

  const networkIdentifier = Buffer.from(
    config.isTestnet ? config.testnetNetworkID : config.networkID,
    'hex'
  );
  if (Platform.OS === 'android') {
    return LiskAndroidPatch.signTransaction(transferAssetSchema, tx, networkIdentifier, passphrase);
  }
  return Lisk.transactions.signTransaction(transferAssetSchema, tx, networkIdentifier, passphrase);
};

export const broadcast = async (transaction) => {
  let txBytes;
  if (Platform.OS === 'android') {
    txBytes = LiskAndroidPatch.getBytes(transaction).toString('hex');
  } else {
    txBytes = Lisk.transactions.getBytes(transferAssetSchema, transaction).toString('hex');
  }
  return apiClient.sendTransaction({ transaction: txBytes });
};
