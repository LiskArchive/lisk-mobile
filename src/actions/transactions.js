import actionTypes from '../constants/actions';
import txConstants from '../constants/transactions';
import { transactions as transactionsAPI } from '../utilities/api';
import { loadingStarted, loadingFinished } from './loading';
import { tokenMap } from '../constants/tokens';

const fetchTransactions = async (dispatch, getState, data) => {
  dispatch(loadingStarted(actionTypes.transactionsLoaded));

  try {
    const response = await transactionsAPI.get(
      getState().settings.token.active,
      data
    );

    dispatch({
      type: actionTypes.transactionsLoaded,
      data: {
        transactions: response.data,
        count: response.meta.count,
      },
    });

    dispatch(loadingFinished(actionTypes.transactionsLoaded));
  } catch (error) {
    dispatch(loadingFinished(actionTypes.transactionsLoaded));
  }
};

export const transactionsReset = () => ({
  type: actionTypes.transactionsReset,
});

export const transactionsLoaded = data => async (dispatch, getState) => {
  fetchTransactions(dispatch, getState, data);
};

/**
 * Calls transactionAPI.create and transactionAPI.broadcast methods to make a transaction.
 * @param {Object} data
 * @param {String} data.recipientAddress
 * @param {Number} data.amount - In raw format (satoshis, beddows)
 * @param {Number} data.fee - In raw format, used for updating the TX List.
 * @param {Number} data.dynamicFeePerByte - In raw format, used for creating BTC transaction.
 * @param {Number} data.reference - Data field for LSK transactions
 * @param {String} data.secondPassphrase - Second passphrase for LSK transactions
 * @param {Function} successCb - success callback
 * @param {Function} errorCb - error callback
 */
export const transactionAdded = (data, successCb, errorCb) => async (
  dispatch,
  getState
) => {
  const {
    settings: { token },
    accounts,
  } = getState();
  const account = accounts.info[token.active];

  try {
    const tx = await transactionsAPI.create(token.active, data);

    if (token.active === tokenMap.LSK.key) {
      const { id } = await transactionsAPI.broadcast(token.active, tx);

      dispatch({
        type: actionTypes.pendingTransactionAdded,
        data: {
          id,
          senderAddress: account.address,
          recipientAddress: data.recipientAddress,
          amount: data.amount,
          fee: data.fee,
          type: txConstants.send.type,
          data: data.reference,
        },
      });

      successCb({ txId: id, address: data.recipientAddress });
    } else {
      await transactionsAPI.broadcast(token.active, tx);

      dispatch(transactionsReset());
      dispatch(transactionsLoaded({ address: account.address }));

      successCb();
    }
  } catch (error) {
    errorCb(error);
  }
};
