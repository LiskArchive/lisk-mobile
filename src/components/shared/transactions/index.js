import React, { Fragment } from 'react';
import {
  View, Animated, Platform, ActivityIndicator
} from 'react-native';
import connect from 'redux-connect-decorator';
import ReactNativeHapticFeedback from 'react-native-haptic-feedback';
import { translate } from 'react-i18next';
import RNShake from 'react-native-shake';
import { settingsUpdated as settingsUpdatedAction } from '../../../actions/settings';
import List from './list';
import Footer from './footer';
import { H3 } from '../toolBox/typography';
import colors from '../../../constants/styleGuide/colors';
import easing from '../../../utilities/easing';
import withTheme from '../withTheme';
import getStyles from './styles';
import { IconButton } from '../toolBox/button';

/**
 * This component is a HOC to decide which state to show:
 * Loading, Empty (No transactions) or the List view.
 *
 * It performs the initial animation after the user logged in.
 */
@connect(
  state => ({
    incognitoMode: state.settings.incognito,
    activeToken: state.settings.token.active,
    followedAccounts: state.accounts.followed || [],
  }),
  {
    settingsUpdated: settingsUpdatedAction,
  }
)
class Transactions extends React.Component {
  state = {
    initialAnimations: {
      opacity: new Animated.Value(0),
      top: new Animated.Value(20),
    },
  };

  componentDidMount() {
    let timeout = null;
    if (this.props.type === 'home') {
      RNShake.addEventListener('ShakeEvent', () => {
        if (!timeout) {
          this.props.settingsUpdated({
            incognito: !this.props.incognitoMode,
          });
          timeout = setTimeout(() => {
            timeout = false;
          }, 1000);
        }
      });
    }
    this.initialFadeIn();
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    RNShake.removeEventListener('ShakeEvent');
  }

  initialFadeIn = () => {
    const { opacity, top } = this.state.initialAnimations;
    Animated.timing(opacity, {
      toValue: 1,
      duration: 400,
      delay: 100,
    }).start();
    Animated.timing(top, {
      toValue: 0,
      duration: 400,
      delay: 100,
      easing: easing.easeInOutQuart,
    }).start();
  };

  toggleIncognito = () => {
    ReactNativeHapticFeedback.trigger('selection');
    this.props.settingsUpdated({
      incognito: !this.props.incognitoMode,
    });
  };

  render() {
    const {
      styles,
      transactions,
      navigate,
      activeToken,
      account,
      footer,
      incognitoMode,
      followedAccounts,
      type,
      t,
    } = this.props;

    const incognito = type === 'home' && incognitoMode;
    const Anim = Animated.View;
    const { opacity, top } = this.state.initialAnimations;

    return (
      <Anim style={[styles.container, { opacity, top }]}>
        {!transactions
          || (transactions.confirmed.length === 0
            && transactions.pending.length === 0) ? (
          <Fragment />
          ) : (
          <Fragment>
            <View style={styles.innerContainer}>
              <H3 style={[styles.title, styles.theme.title]}>
                {t('Activity')}
              </H3>
            </View>
            <List
              incognito={incognito}
              navigate={navigate}
              account={account ? account.address : ''}
              followedAccounts={followedAccounts}
              pending={transactions.pending}
              activeToken={activeToken}
                transactions={[...transactions.confirmed, ...transactions.confirmed, ...transactions.confirmed, ...transactions.confirmed]}
            />
            {footer ? <Footer /> : null}
          </Fragment>
          )}
      </Anim>
    );
  }
}
export default withTheme(translate()(Transactions), getStyles());
