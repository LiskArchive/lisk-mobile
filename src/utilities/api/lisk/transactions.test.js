import { apiClient } from './apiClient';
import * as transactions from './transactions';

jest.mock('./apiClient');

const response = {
  meta: { offset: 0, limit: 10, total: 2 },
  data: [
    {
      id: '10703320139793065364',
      height: 7450256,
      blockId: '9937261865542489722',
      type: 0,
      timestamp: 85447190,
      senderPublicKey: '84795ae68a1908d2a89daafce3f9980811ddf8f0c087c913ebe57759c977fbbc',
      recipientPublicKey: '04ffd6fb4508a5ca85355cc3b30dce326fe35781d32deb858879128cf8c141cc',
      amount: '1000000',
      fee: '10000000',
      signature:
        '08d2526bcd49299bb6894fb62a39b5d084c132cb593be674096aad8b273190922f24677def02817b424c7db1130f14799a0ffc4f95744980f2120c0f1376e200',
      signatures: [],
      confirmations: 99170,
      asset: {},
      senderId: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
      recipientId: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
      sender: {
        address: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
      },
      recipient: {
        address: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
      },
      block: {
        id: '381b0cc8ced666899b8113d3b4837cd3314bb440c5d84a6eba5f027992f9ac13',
        timestamp: 72473846,
        height: 15201096
      }
    },
    {
      id: '17545660133359630206',
      height: 7516990,
      blockId: '7795918980027524180',
      type: 0,
      timestamp: 86128728,
      senderPublicKey: '04ffd6fb4508a5ca85355cc3b30dce326fe35781d32deb858879128cf8c141cc',
      recipientPublicKey: '',
      amount: '10000000',
      fee: '10000000',
      signature:
        'b2eb00349cab3158076f29eab8f83250828ffafaee0439f3666236af9dff4d51949652cd155beb5f3d93a836b6f375e730de0f16fb4c659c7c67ed67b0849f0b',
      signatures: [],
      confirmations: 32436,
      asset: {
        data: 'test'
      },
      senderId: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
      recipientId: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
      sender: {
        address: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
      },
      recipient: {
        address: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
      },
      block: {
        id: '381b0cc8ced666899b8113d3b4837cd3314bb440c5d84a6eba5f027992f9ac13',
        timestamp: 72473846,
        height: 15101096
      }
    }
  ],
  links: {}
};

describe.skip('api/transactions', () => {
  beforeAll(() => {
    apiClient.getTransactions = jest.fn();
    apiClient.getTransactions.mockResolvedValue({ meta: response.meta, data: response.data });
    apiClient.sendTransaction = jest.fn();
    apiClient.getLatestBlock = jest.fn();
    apiClient.getLatestBlock.mockResolvedValue({
      height: 15301076,
      timestamp: 1640594580
    });
  });

  describe('get', () => {
    it('should resolve correctly', async () => {
      const result = await transactions.get({
        address: response.data[0].recipient.address
      });
      expect(apiClient.getTransactions).toBeCalledWith(
        response.data[0].recipient.address,
        undefined,
        undefined
      );
      expect(result).toEqual({
        data: [
          {
            id: '10703320139793065364',
            senderAddress: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
            recipientAddress: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
            amount: 0,
            fee: '10000000',
            timestamp: 72473846,
            confirmations: 99981,
            nonce: undefined,
            type: undefined,
            moduleAssetId: undefined,
            moduleAssetName: undefined,
            blockHeight: 15201096,
            blockId: '381b0cc8ced666899b8113d3b4837cd3314bb440c5d84a6eba5f027992f9ac13',
            data: '',
            votes: [],
            delegate: ''
          },
          {
            id: '17545660133359630206',
            senderAddress: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
            recipientAddress: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw',
            amount: 0,
            fee: '10000000',
            timestamp: 72473846,
            confirmations: 199981,
            nonce: undefined,
            type: undefined,
            moduleAssetId: undefined,
            moduleAssetName: undefined,
            blockHeight: 15101096,
            blockId: '381b0cc8ced666899b8113d3b4837cd3314bb440c5d84a6eba5f027992f9ac13',
            data: '',
            votes: [],
            delegate: ''
          }
        ],
        meta: { limit: undefined, offset: undefined, count: 2 }
      });
    });
    describe('empty account', () => {
      beforeAll(() => {
        apiClient.getTransactions.mockResolvedValue({ meta: undefined, data: undefined });
      });
      it('should resolve correctly for empty accounts', async () => {
        const result = await transactions.get({
          address: response.data[0].recipient.address
        });
        expect(apiClient.getTransactions).toBeCalledWith(
          response.data[0].recipient.address,
          undefined,
          undefined
        );
        expect(result).toEqual({
          data: [],
          meta: {}
        });
      });

      it('Should handle rejections', async () => {
        const errorMessage = { message: 'Error!' };
        apiClient.getTransactions.mockRejectedValueOnce(errorMessage);

        try {
          await transactions.get({ address: response.data[0].recipientId });
        } catch (error) {
          expect(error).toBe(errorMessage);
        }
      });
    });
  });

  it('Should call transactions.create method correctly', async () => {
    const tx = await transactions.create({
      nonce: '5',
      passphrase: 'test',
      recipientAddress: response.data[0].recipientId,
      amount: response.data[0].amount,
      fee: '100'
    });
    expect(tx).toMatchSnapshot();
  });

  it('Should call transactions.broadcast method correctly', async () => {
    apiClient.sendTransaction.mockResolvedValueOnce(response.data[0]);
    const tx = await transactions.create({
      nonce: '6',
      passphrase: 'test',
      recipientAddress: response.data[0].recipientId,
      amount: response.data[0].amount,
      fee: '100'
    });
    const result = await transactions.broadcast(tx);
    expect(result).toMatchSnapshot();
  });

  it('Should handle rejections from transactions.broadcast method', async () => {
    const errorMessage = { message: 'Error!' };
    apiClient.sendTransaction.mockRejectedValueOnce(errorMessage);

    try {
      await transactions.broadcast(response.data[0]);
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe.skip('Should get transaction detail for one transaction', () => {
  beforeAll(() => {
    apiClient.getTransaction = jest.fn();
    apiClient.getTransaction.mockResolvedValue([
      {
        ...response.data[0],
        moduleAssetId: '2:0',
        asset: {
          amount: '100000',
          recipient: {
            address: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
          }
        }
      }
    ]);
  });

  it('Should call get transaction with id to get one transaction detail', async () => {
    await transactions.get({
      id: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
    });
    expect(apiClient.getTransaction).toBeCalledWith('lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw');
  });

  it('Should extract amount for transfer type', async () => {
    const result = await transactions.get({
      id: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
    });
    expect(result.data[0].amount).toEqual('100000');
  });
});

describe.skip('Should get amount from unlock token transaction type', () => {
  beforeAll(() => {
    apiClient.getTransaction = jest.fn();
    apiClient.getLatestBlock = jest.fn();
    apiClient.getLatestBlock.mockResolvedValue({
      height: 15301076,
      timestamp: 1640594580
    });
    apiClient.getTransaction.mockResolvedValue([
      {
        ...response.data[0],
        moduleAssetId: '5:2',
        asset: {
          amount: '100000',
          recipient: {
            address: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
          },
          unlockObjects: [{ amount: 10 }, { amount: 11 }, { amount: 100 }, { amount: 98 }]
        }
      }
    ]);
  });

  it('should extract amount for unlock token type', async () => {
    const result = await transactions.get({
      id: 'lskebd9zfkhz6ep9kde24u8h7uxarssxxdnru2xgw'
    });
    expect(result.data[0].amount).toEqual('219');
  });
});
