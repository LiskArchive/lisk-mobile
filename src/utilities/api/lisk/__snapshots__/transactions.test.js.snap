// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/transactions Should handle rejections from transactions.broadcast method 1`] = `[TypeError: Cannot read properties of undefined (reading 'length')]`;
